### PR TITLE
feat: add standalone SQLite backed denokv server

### PR DESCRIPTION
**Pull request by lucacasonato**
This is a standalone Deno KV server backed by SQLite. It exposes a HTTP API that
implements the KV Connect protocol and can be used from the Deno CLI.

The server can be connected to from multiple clients at the same time, which
enables a simple way to self host a production ready Deno KV database.

### CODE REVIEW

File: denokv/main.rs [hunk 1]
```diff
@@ -0,0 +1,584 @@
...
+      return Err(ApiError::KeyTooLong);
+    }
+    total_payload_size += mutation.key.len();
+    let value_size = mutation.value.as_ref().map(|v| v.data.len()).unwrap_or(0);
+    if value_size > limits::MAX_VALUE_SIZE_BYTES {
+      return Err(ApiError::ValueTooLong);
+    }
+    total_payload_size += value_size;
+
+    let kind = match (mutation.mutation_type, mutation.value) {
```
Comment (losfair) [inline denokv/main.rs:313]: match on `mutation.mutation_type()` to use the enum variants

File: denokv/main.rs [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+  let Some((bearer, token)) = authorization.split_once(' ') else {
+    return Err(ApiError::MalformedAuthorizationHeader);
+  };
+  if bearer.to_lowercase() != "bearer" || token != state.access_token {
+    return Err(ApiError::InvalidAccessToken);
+  }
+  let expires_at = utc_now() + Duration::days(1);
+  Ok(Json(DatabaseMetadata {
+    version: 2,
+    database_id: Uuid::nil(),
```
Comment (losfair) [inline denokv/main.rs:138]:
```rust
static MAGIC_UUID: &'static str = some_pre_generated_random_uuid;
```

?

File: denokv/main.rs [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+  }
+  let Some(authorization) =
+    headers.get("authorization").and_then(|v| v.to_str().ok())
+  else {
+    return Err(ApiError::MalformedAuthorizationHeader);
+  };
+  let Some((bearer, token)) = authorization.split_once(' ') else {
+    return Err(ApiError::MalformedAuthorizationHeader);
+  };
+  if bearer.to_lowercase() != "bearer" || token != state.access_token {
```
Comment (losfair) [inline denokv/main.rs:132]: constant time eq

File: denokv/main.rs (hunk 2 above)
Comment (lucacasonato) [reply denokv/main.rs:138]: Why?

File: denokv/main.rs [hunk 4]
```diff
@@ -0,0 +1,584 @@
...
+    return Err(ApiError::InvalidAccessToken);
+  }
+  let expires_at = utc_now() + Duration::days(1);
+  Ok(Json(DatabaseMetadata {
+    version: 2,
+    database_id: Uuid::nil(),
+    endpoints: vec![EndpointInfo {
```
Comment (losfair) [inline denokv/main.rs:139]: Advertise Alt-Svc?

File: denokv/main.rs (hunk 2 above)
Comment (losfair) [reply denokv/main.rs:138]: or maybe a randomly generated UUID, the zero UUID can have special semantics somewhere

File: denokv/main.rs (hunk 4 above)
Comment (lucacasonato) [reply denokv/main.rs:139]: For what? h3?

Comment (losfair) [reply denokv/main.rs:139]: Yes, and turn on `http2_prior_knowledge` by default

File: denokv/main.rs [hunk 5]
```diff
@@ -0,0 +1,584 @@
...
+      url: Cow::Borrowed("/v1"),
```
Comment (igorzi) [inline denokv/main.rs:140]: nit: should this be `/v2` (to avoid confusion if someone looks at the url)

File: denokv/main.rs (hunk 2 above)
Comment (lucacasonato) [reply denokv/main.rs:138]: But a random UUID changes across restarts, which seems bad

Comment (losfair) [reply denokv/main.rs:138]: Ok, let's use `Uuid::nil()` for now - but I think we should assign an ID to SQLite KV databases in another patch

File: denokv/main.rs (hunk 4 above)
Comment (lucacasonato) [reply denokv/main.rs:146]: So you want to add an `altSvc` entry to `EndpointInfo`?

File: README.md [hunk 6]
```diff
@@ -1,5 +1,168 @@
...
+of your VPS in the URL passed to `Deno.openKv`.
+
+You should additionally add a HTTPS terminating proxy or loadbalancer in front
+of `denokv` to ensure that all communication happens over TLS. Not using TLS can
+pose a significant security risk. The HTTP protocol used by Deno KV is
+compatible with any HTTP proxy, such as `caddy`, `nginx`, or a loadbalancer.
+
+### Fly.io
+
+You can easially host `denokv` on https://fly.io.
```
Comment (losfair) [inline README.md:86]: typo

File: denokv/main.rs (hunk 4 above)
Comment (losfair) [reply denokv/main.rs:146]: Either that or we should support DNS SVCB/HTTPS records

Comment (losfair) [reply denokv/main.rs:146]: This can be another patch though - it's a protocol change

**Review by losfair — APPROVED** [review]
LGTM

**Pull request merged**